datafusion/optimizer/src/rewrite_disjunctive_predicate.rs

```diff
@@ -0,0 +1,468 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::{OptimizerConfig, OptimizerRule};
+use datafusion_common::Result;
+use datafusion_expr::logical_plan::{
+    Aggregate, Analyze, CreateMemoryTable, CreateView, CrossJoin, Distinct, Explain,
+    Filter, Join, Limit, Projection, Repartition, Sort, Subquery, SubqueryAlias, Union,
+    Window,
+};
+use datafusion_expr::Expr::BinaryExpr;
+use datafusion_expr::{Expr, LogicalPlan, Operator};
+use std::sync::Arc;
+
+#[derive(Clone, PartialEq, Debug)]
+enum Predicate {
+    And { args: Vec<Predicate> },
+    Or { args: Vec<Predicate> },
+    Other { expr: Box<Expr> },
+}
+
+fn predicate(expr: &Expr) -> Result<Predicate> {
```

Comment:
I wonder if using `ExprRewriter` here might make the code less verbose (and avoid a copy). Doing so would require a non trivial refactor though

Reply:
The method is to recursely extract all args of `And` or `Or`, such as `a > 1 and b and c + 1 = 2` => `Predicate::And {[a > 1, b, c + 1 =2]}`. The return value is `Predicate` not `Expr`, so seems can't reuse `ExprRewriter`?